File: LeetCode/python3/257.py
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root:
            return []
        ret = []
        self.BackTrack(root, str(root.val), ret)
        return ret

    def BackTrack(self, node, s, ret):
        if not node.left and not node.right:
            ret.append(s)
            return
        if node.left:
            self.BackTrack(node.left, s + "->" + str(node.left.val), ret)
        if node.right:
            self.BackTrack(node.right, s + "->" + str(node.right.val), ret)

File: LeetCode/cpp/257.cpp
class Solution {
public:
    vector<string> ret;

    vector<string> binaryTreePaths(TreeNode *root) {
        ret.clear();
        if (!root)
            return ret;
        string path;
        path += to_string(root->val);
        helper(root->left, path);
        helper(root->right, path);
        if (ret.empty())
            ret.push_back(path);
        return ret;
    }

    void helper(TreeNode *root, string &path) {
        if (!root)
            return;
        if (!root->left && !root->right) {
            string &&temp = path + "->" + to_string(root->val);
            ret.push_back(temp);
            return;
        }
        int &&size = path.size();
        path += "->" + to_string(root->val);
        helper(root->left, path);
        helper(root->right, path);
        path = path.substr(0, size);
    }
};